File: platform/lang-api/src/com/intellij/execution/runners/ExecutionEnvironmentBuilder.java
/*
 * Copyright 2000-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intellij.execution.runners;

import com.intellij.execution.*;
import com.intellij.execution.configurations.ConfigurationPerRunnerSettings;
import com.intellij.execution.configurations.RunConfiguration;
import com.intellij.execution.configurations.RunProfile;
import com.intellij.execution.configurations.RunnerSettings;
import com.intellij.execution.ui.RunContentDescriptor;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.UserDataHolderBase;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class ExecutionEnvironmentBuilder {
  @NotNull private RunProfile myRunProfile;
  @NotNull private ExecutionTarget myTarget = DefaultExecutionTarget.INSTANCE;

  @NotNull private final Project myProject;

  @Nullable private RunnerSettings myRunnerSettings;
  @Nullable private ConfigurationPerRunnerSettings myConfigurationSettings;
  @Nullable private RunContentDescriptor myContentToReuse;
  @Nullable private RunnerAndConfigurationSettings myRunnerAndConfigurationSettings;
  @Nullable private String myRunnerId;
  private ProgramRunner<?> myRunner;
  private boolean myAssignNewId;
  @NotNull private Executor myExecutor;
  @Nullable private DataContext myDataContext;
  private final UserDataHolderBase myUserData = new UserDataHolderBase();

  public ExecutionEnvironmentBuilder(@NotNull Project project, @NotNull Executor executor) {
    myProject = project;
    myExecutor = executor;
  }

  @NotNull
  public static ExecutionEnvironmentBuilder create(@NotNull Project project, @NotNull Executor executor, @NotNull RunProfile runProfile) throws ExecutionException {
    ExecutionEnvironmentBuilder builder = createOrNull(project, executor, runProfile);
    if (builder == null) {
      throw new ExecutionException("Cannot find runner for " + runProfile.getName());
    }
    return builder;
  }

  @Nullable
  public static ExecutionEnvironmentBuilder createOrNull(@NotNull Project project, @NotNull Executor executor, @NotNull RunProfile runProfile) {
    ProgramRunner runner = RunnerRegistry.getInstance().getRunner(executor.getId(), runProfile);
    if (runner == null) {
      return null;
    }
    return new ExecutionEnvironmentBuilder(project, executor).runner(runner).runProfile(runProfile);
  }

  @Nullable
  public static ExecutionEnvironmentBuilder createOrNull(@NotNull Executor executor, @NotNull RunnerAndConfigurationSettings settings) {
    ExecutionEnvironmentBuilder builder = createOrNull(settings.getConfiguration().getProject(), executor, settings.getConfiguration());
    return builder == null ? null : builder.runnerAndSettings(builder.myRunner, settings);
  }

  @NotNull
  public static ExecutionEnvironmentBuilder create(@NotNull Executor executor, @NotNull RunnerAndConfigurationSettings settings) throws ExecutionException {
    RunConfiguration configuration = settings.getConfiguration();
    ExecutionEnvironmentBuilder builder = create(configuration.getProject(), executor, configuration);
    return builder.runnerAndSettings(builder.myRunner, settings);
  }

  @NotNull
  public static ExecutionEnvironmentBuilder create(@NotNull Executor executor, @NotNull RunConfiguration configuration) {
    return new ExecutionEnvironmentBuilder(configuration.getProject(), executor).runProfile(configuration);
  }

  @NotNull
  Executor getExecutor() {
    return myExecutor;
  }

  /**
   * Creates an execution environment builder initialized with a copy of the specified environment.
   *
   * @param copySource the environment to copy from.
   */
  public ExecutionEnvironmentBuilder(@NotNull ExecutionEnvironment copySource) {
    myTarget = copySource.getExecutionTarget();
    myProject = copySource.getProject();
    myRunnerAndConfigurationSettings = copySource.getRunnerAndConfigurationSettings();
    myRunProfile = copySource.getRunProfile();
    myRunnerSettings = copySource.getRunnerSettings();
    myConfigurationSettings = copySource.getConfigurationSettings();
    //noinspection deprecation
    myRunner = copySource.getRunner();
    myContentToReuse = copySource.getContentToReuse();
    myExecutor = copySource.getExecutor();
    copySource.copyUserDataTo(myUserData);
  }

  public ExecutionEnvironmentBuilder target(@Nullable ExecutionTarget target) {
    if (target != null) {
      myTarget = target;
    }
    return this;
  }

  public ExecutionEnvironmentBuilder activeTarget() {
    myTarget = ExecutionTargetManager.getActiveTarget(myProject);
    return this;
  }

  public ExecutionEnvironmentBuilder runnerAndSettings(@NotNull ProgramRunner runner,
                                                       @NotNull RunnerAndConfigurationSettings settings) {
    myRunnerAndConfigurationSettings = settings;
    myRunProfile = settings.getConfiguration();
    myRunnerSettings = settings.getRunnerSettings(runner);
    myConfigurationSettings = settings.getConfigurationSettings(runner);
    myRunner = runner;
    return this;
  }

  public ExecutionEnvironmentBuilder runnerSettings(@Nullable RunnerSettings runnerSettings) {
    myRunnerSettings = runnerSettings;
    return this;
  }

  public ExecutionEnvironmentBuilder contentToReuse(@Nullable RunContentDescriptor contentToReuse) {
    myContentToReuse = contentToReuse;
    return this;
  }

  public ExecutionEnvironmentBuilder runProfile(@NotNull RunProfile runProfile) {
    myRunProfile = runProfile;
    return this;
  }

  public ExecutionEnvironmentBuilder runner(@NotNull ProgramRunner<?> runner) {
    myRunner = runner;
    return this;
  }

  public ExecutionEnvironmentBuilder dataContext(@Nullable DataContext dataContext) {
    myDataContext = dataContext;
    return this;
  }

  public ExecutionEnvironmentBuilder executor(@NotNull Executor executor) {
    myExecutor = executor;
    return this;
  }

  @NotNull
  public ExecutionEnvironment build() {
    ExecutionEnvironment environment = null;
    ExecutionEnvironmentProvider environmentProvider = ServiceManager.getService(myProject, ExecutionEnvironmentProvider.class);
    if (environmentProvider != null) {
      environment = environmentProvider.createExecutionEnvironment(
        myProject, myRunProfile, myExecutor, myTarget, myRunnerSettings, myConfigurationSettings, myRunnerAndConfigurationSettings);
    }

    if (environment == null && myRunner == null) {
      if (myRunnerId == null) {
        myRunner = RunnerRegistry.getInstance().getRunner(myExecutor.getId(), myRunProfile);
      }
      else {
        myRunner = RunnerRegistry.getInstance().findRunnerById(myRunnerId);
      }
    }

    if (environment == null && myRunner == null) {
      throw new IllegalStateException("Runner must be specified");
    }

    if (environment == null) {
      environment = new ExecutionEnvironment(myRunProfile, myExecutor, myTarget, myProject, myRunnerSettings,
                                             myConfigurationSettings, myContentToReuse, myRunnerAndConfigurationSettings, myRunner);
    }

    if (myAssignNewId) {
      environment.assignNewExecutionId();
    }
    if (myDataContext != null) {
      environment.setDataContext(myDataContext);
    }
    myUserData.copyUserDataTo(environment);
    return environment;
  }

  public void buildAndExecute() throws ExecutionException {
    ExecutionEnvironment environment = build();
    myRunner.execute(environment);
  }
}

File: platform/execution/src/com/intellij/execution/runners/ExecutionEnvironmentBuilder.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.execution.runners

import com.intellij.execution.*
import com.intellij.execution.configurations.ConfigurationPerRunnerSettings
import com.intellij.execution.configurations.RunConfiguration
import com.intellij.execution.configurations.RunProfile
import com.intellij.execution.configurations.RunnerSettings
import com.intellij.execution.ui.RunContentDescriptor
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.UserDataHolderBase

class ExecutionEnvironmentBuilder(private val project: Project, private var executor: Executor) {
  private var runProfile: RunProfile? = null
  private var target = DefaultExecutionTarget.INSTANCE
  private var runnerSettings: RunnerSettings? = null
  private var configurationSettings: ConfigurationPerRunnerSettings? = null
  private var contentToReuse: RunContentDescriptor? = null
  private var runnerAndConfigurationSettings: RunnerAndConfigurationSettings? = null
  private var runner: ProgramRunner<*>? = null
  private var executionId: Long? = null
  private var dataContext: DataContext? = null
  private val userData = UserDataHolderBase()
  private var modulePath: String? = null

  /**
   * Creates an execution environment builder initialized with a copy of the specified environment.
   *
   * @param env the environment to copy from.
   */
  constructor(env: ExecutionEnvironment) : this(env.project, env.executor) {
    target = env.executionTarget
    runnerAndConfigurationSettings = env.runnerAndConfigurationSettings
    runProfile = env.runProfile
    runnerSettings = env.runnerSettings
    configurationSettings = env.configurationSettings
    runner = env.runner
    contentToReuse = env.contentToReuse
    env.copyUserDataTo(userData)
  }

  companion object {
    @JvmStatic
    @Throws(ExecutionException::class)
    fun create(project: Project, executor: Executor, runProfile: RunProfile): ExecutionEnvironmentBuilder {
      return createOrNull(project, executor, runProfile)
             ?: throw ExecutionException(ExecutionBundle.message("dialog.message.cannot.find.runner", runProfile.name))
    }

    @JvmStatic
    fun createOrNull(project: Project, executor: Executor, runProfile: RunProfile): ExecutionEnvironmentBuilder? {
      val runner = ProgramRunner.getRunner(executor.id, runProfile) ?: return null
      return ExecutionEnvironmentBuilder(project, executor).runner(runner).runProfile(runProfile)
    }

    @JvmStatic
    fun createOrNull(executor: Executor, settings: RunnerAndConfigurationSettings): ExecutionEnvironmentBuilder? {
      val builder = createOrNull(executor, settings.configuration)
      return builder?.runnerAndSettings(builder.runner!!, settings)
    }

    @JvmStatic
    fun createOrNull(executor: Executor, configuration: RunConfiguration): ExecutionEnvironmentBuilder? {
      val builder = createOrNull(configuration.project, executor, configuration)
      builder?.runProfile(configuration)
      return builder
    }

    @JvmStatic
    @Throws(ExecutionException::class)
    fun create(executor: Executor, settings: RunnerAndConfigurationSettings): ExecutionEnvironmentBuilder {
      val configuration = settings.configuration
      val builder = create(configuration.project, executor, configuration)
      return builder.runnerAndSettings(builder.runner!!, settings)
    }

    @JvmStatic
    fun create(executor: Executor, configuration: RunConfiguration): ExecutionEnvironmentBuilder {
      return ExecutionEnvironmentBuilder(configuration.project, executor).runProfile(configuration)
    }
  }

  fun target(target: ExecutionTarget?): ExecutionEnvironmentBuilder {
    if (target != null) {
      this.target = target
    }
    return this
  }

  fun activeTarget(): ExecutionEnvironmentBuilder {
    target = ExecutionTargetManager.getActiveTarget(project)
    return this
  }

  fun runnerAndSettings(runner: ProgramRunner<*>, settings: RunnerAndConfigurationSettings): ExecutionEnvironmentBuilder {
    runnerAndConfigurationSettings = settings
    runProfile = settings.configuration
    runnerSettings = settings.getRunnerSettings(runner)
    configurationSettings = settings.getConfigurationSettings(runner)
    this.runner = runner
    return this
  }

  fun runnerSettings(runnerSettings: RunnerSettings?): ExecutionEnvironmentBuilder {
    this.runnerSettings = runnerSettings
    return this
  }

  fun contentToReuse(contentToReuse: RunContentDescriptor?): ExecutionEnvironmentBuilder {
    this.contentToReuse = contentToReuse
    return this
  }

  fun runProfile(runProfile: RunProfile): ExecutionEnvironmentBuilder {
    this.runProfile = runProfile
    return this
  }

  fun runner(runner: ProgramRunner<*>): ExecutionEnvironmentBuilder {
    this.runner = runner
    return this
  }

  fun dataContext(dataContext: DataContext?): ExecutionEnvironmentBuilder {
    this.dataContext = dataContext
    return this
  }

  fun executor(executor: Executor): ExecutionEnvironmentBuilder {
    this.executor = executor
    return this
  }

  fun executionId(executionId: Long): ExecutionEnvironmentBuilder {
    this.executionId = executionId
    return this
  }

  fun modulePath(modulePath: String): ExecutionEnvironmentBuilder {
    this.modulePath = modulePath
    return this
  }

  @JvmOverloads
  fun build(callback: ProgramRunner.Callback? = null): ExecutionEnvironment {
    var environment: ExecutionEnvironment? = null
    val environmentProvider = project.getService(ExecutionEnvironmentProvider::class.java)
    if (environmentProvider != null) {
      environment = environmentProvider.createExecutionEnvironment(
        project, runProfile!!, executor, target, runnerSettings, configurationSettings, runnerAndConfigurationSettings)
    }
    if (environment == null && runner == null) {
      runner = ProgramRunner.getRunner(executor.id, runProfile!!)
    }
    if (environment == null && runner == null) {
      throw IllegalStateException("Runner must be specified")
    }
    if (environment == null) {
      environment = ExecutionEnvironment(runProfile!!, executor, target, project, runnerSettings,
                                         configurationSettings, contentToReuse,
                                         runnerAndConfigurationSettings, runner!!, callback)
    }
    if (executionId != null) {
      environment.executionId = executionId!!
    }
    if (dataContext != null) {
      environment.setDataContext(dataContext!!)
    }
    if (modulePath != null) {
      environment.setModulePath(modulePath!!)
    }
    userData.copyUserDataTo(environment)
    return environment
  }

  @Throws(ExecutionException::class)
  fun buildAndExecute() {
    val environment = build()
    runner!!.execute(environment)
  }
}